### PR TITLE
FLINK-12751#File System#Implementation of HA using file system

### PR DESCRIPTION
**Pull request by blublinsky**
## What is the purpose of the change

This pull request implements Flink HA on file system - no ZK required

## Brief change log

An implementation overwrites 2 of the Flink files:
* HighAvailabilityServicesUtils - added `FILESYSTEM` option for picking HA service
* HighAvailabilityMode - added `FILESYSTEM` to available HA options.

The actual implementation exists of the following classes:
* FileSystemHAServices - an implementation of a HighAvailabilityServices for file system
* FileSystemUtils - support class for creation of runtime components.
* FileSystemStorageHelper - file system operations implementation for filesystem based HA
* FileSystemCheckpointRecoveryFactory - an implementation of a CheckpointRecoveryFactory for file system
* FileSystemCheckpointIDCounter - an implementation of a CheckpointIDCounter for file system
* FileSystemCompletedCheckpointStore - an implementation of a CompletedCheckpointStore for file system
* FileSystemSubmittedJobGraphStore - an implementation of a SubmittedJobGraphStore for file system

## Verifying this change

Here we are using Flink's native leader election, which is fully tested by Flink. We also implement 
Flink interfaces, so connectivity between pieces is also tested by Flink.

Also manually verified the change by running a 2 node cluser with 1 JobManagers and 2 TaskManagers, a stateful streaming program, and killing one JobManager during the execution, verifying that recovery happens correctly

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: yes - it implements a new option for JM checkpointing and recovery
  - The S3 file system connector: no
## Documentation

  - Does this pull request introduce a new feature? yes 
  - If yes, how is the feature documented? documented in code
This is how data is laid out on file system
 ha									-----> root of the HA data
  checkpointcounter					-----> checkpoint counter folder
  	<job ID>						-----> job id folder
  		<counter file>				-----> counter file
  	<another job ID>				-----> another job id folder
  ...........
  completedCheckpoint				-----> completed checkpoint folder
 	<job ID>						-----> job id folder
  		<checkpoint file>			-----> checkpoint file
  		<another checkpoint file>	-----> checkpoint file
 		...........
  		<another job ID>			-----> another job id folder
  ...........
  submittedJobGraph				    -----> submitted graph folder
  	<job ID>						-----> job id folder
  		<graph file>				-----> graph file
  	<another job ID>				-----> another job id folder
  ...........

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/filesystem/FileSystemHAServices.java [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+			return miniclusterURLResolver.getResourceManagerLeaderRetriever();
+		}
+		else {
+			return clusterURLResolver.getResourceManagerLeaderRetriever();
+		}
+	}
+
+	@Override
+	public LeaderRetrievalService getDispatcherLeaderRetriever() {
+		if(local) {
```
Comment (JTaky) [inline flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/filesystem/FileSystemHAServices.java]: IMHO polymorphism would make this code easier to read

File: flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/filesystem/FileSystemUtils.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    private static <T extends Serializable> FileSystemStorageHelper<T> createFileSystemStateStorage(
+            Configuration configuration,
+            String prefix) throws IOException {
+
+        String storagePath = configuration.getValue(HighAvailabilityOptions.HA_STORAGE_PATH) + "/ha";
+        if(storagePath.startsWith("file://")) {
+            storagePath = storagePath.substring(7);
+        }
+        LOG.info("Creating FileSystemStorageHelper with path " + storagePath + " prefix " + prefix);
```
Comment (JTaky) [inline flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/filesystem/FileSystemUtils.java]: LOG.info("Creating FileSystemStorageHelper with path {} prefix {}", storagePath, prefix)

File: flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/filesystem/FileSystemHAServices.java (hunk 1 above)
Comment (blublinsky) [reply flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/filesystem/FileSystemHAServices.java]: Fixing it now. Missed the common parent

**Comment by klion26** [comment]
thanks for the contribution @blublinsky. Could you please rewrite the pr title such as `[FLINK-12757][${component}][${PR title}]`, so that the others can easily know what does this pr do from the title.

**Comment by tillrohrmann** [comment]
Closing this PR since it looks to be abandoned.

**Pull request closed without merging**